### ISSUE TITLE
Auto Saving & Offline Support

### ISSUE THREAD

**Issue by bootstraponline** [gollum/gollum#527]
As mentioned on #525.

It'd be easy to use HTML5 local storage to automatically save posts and provide offline support. Then if the browser crashed or the internet disconnected, no work would be lost.

**Comment by bartkamphorst** [comment]
Here's a straightforward tutorial on [how to use localStorage](https://toddmotto.com/storing-data-in-the-browser-with-the-html5-local-storage-api/). @dometto I think this shouldn't be too difficult to implement. I'd like to discuss what the correct behavior should be though. Here's a first proposal:

* Creating or editing a page means storing page content data to localStorage. For keys, we could store page data with `'gollum-' + page.sha` as the key while editing an existing page, and `gollum-new-page` whilst creating a new page. 
* When saving the page, we should delete the page content from localStorage. 
* When hitting 'Cancel' the localStorage should also be emptied.
* When hitting the 'Back' button, the localStorage should persist.
* When creating or editing a page gollum should look if the browser has any stored data, and if so, overwrite the loaded data from the server with the stored data from the browser. This is the scenario where one is trying to recover from a crash or accidental closing of a tab.
* When a version conflict is detected, gollum should empty the localStorage, since a reload of the page will give a different `sha` and we don't want to keep increasing storage space.

Questions:
* Should gollum alert the user about recovering stored data or should it "just work"?
* How often should autosave run? Every second?

**Comment by dometto** [comment]
On second thought, maybe it would be nice if there was an unobtrusive text, say just above or below the Save button (or wherever), 'Gollum just recovered your changes (undo)' (where clicking undo reloads the page's content)?

**Comment by pedrocr** [comment]
A probably simple sub-feature would be to just have a way for hitting save in the edit view and have that just save but not go back to the normal view. That way I can hit save just like I do in my normal text-editor, to make sure the changes made it to disk, while continuing to edit.

**Comment by mivok** [comment]
If anyone is interested in having something like this now, I have made the following in my custom.js file which implements autosaving using local storage. It probably needs some cleanup/style changes to be usable in gollum (e.g. this uses some newer JS features, and doesn't make use of jquery, which I believe gollum does, and I haven't made any attempt to make it work across multiple browsers), but it works well for me:

```js
//
// Autorecover/autosave functionality when editing
//
function autoRecover() {
    // Don't do anything if there isn't an editor
    if (!window.ace_editor) { return }

    const storageKey = `autorecover_${window.location}`;
    const savedValue = window.localStorage.getItem(storageKey);

    // Keep track of the timer on change before autosaving
    let autoSaveTimer = null;

    // If we have any saved text, then show a prompt to restore it
    if (savedValue) {
        const savedMsg = document.createElement('div');
        savedMsg.className = 'flash flash-full'
        savedMsg.innerHTML = `
        Autosaved text is available. Click the button to restore it.
        <button class="btn btn-sm primary flash-action">
            Restore text
        </button>
        `;
        const restoreButton = savedMsg.querySelector('button');
        restoreButton.addEventListener('click', () => {
            window.ace_editor.setValue(savedValue);
            savedMsg.remove();
        });
        const bodyElem = document.querySelector('body')
        bodyElem.insertBefore(savedMsg, bodyElem.firstChild);
    }

    // Set up a message element
    const messageElem = document.createElement('div');
    messageElem.id = 'autorecover-msg';
    messageElem.className = 'div position-fixed bottom-0 left-0';
    document.querySelector('body').appendChild(messageElem);

    // Actually do the autosaving after every change (but wait a few seconds
    // first)
    window.ace_editor.on('change', () => {
        if (autoSaveTimer) {
            clearTimeout(autoSaveTimer)
        }

        // Show Saving... when waiting before saving
        messageElem.innerHTML = 'Saving...';

        autoSaveTimer = setTimeout(() => {
            window.localStorage.setItem(storageKey,
                window.ace_editor.getValue());
            messageElem.innerHTML = 'Saved recovery text';
        }, 2000);
    });

    // Clear any autosaved values on save or cancel
    [
        '#gollum-editor-submit',
        '#gollum-editor-cancel'
    ].forEach((selector) => {
        document.querySelector(selector).addEventListener(
            'click', () => {
                window.localStorage.removeItem(storageKey);
            }
        );
    });
}
document.addEventListener('DOMContentLoaded', autoRecover, false);
```

There are a few differences in behavior from what is described above:

* The key for saving is based on the current URL, and not on a SHA
* You are prompted at the top of the page if there is text to recover. That way you don't have to load the saved text if it's not appropriate.
* Autosaving happens 2 seconds after you stop typing/making changes to the document

If this seems like a decent approach, I can make an attempt to clean it up into something more suitable for inclusion in gollum itself and open a PR.

**Comment by bartkamphorst** [comment]
@mivok Thanks for sharing your solution. I would very much welcome a PR implementing this in gollum! As it happens, this is currently the oldest open request in gollum, so it would be terrific to add this functionality to 5.1. RE the key differences you mention: 

> The key for saving is based on the current URL, and not on a SHA

Seems sensible. @dometto do you see any reason to prefer a SHA here?

> You are prompted at the top of the page if there is text to recover. That way you don't have to load the saved text if it's not appropriate.

My sense is that it would be more intuitive to recover the text by default (and maybe add an undo button). That is how Github does it, and I'm always reassured when I return to a text field I was editing after accidentally hitting the back button or closing my active tab and my text is just there. Intuitions may diverge here, though. @dometto @heavywatal care to weigh in on this?

> Autosaving happens 2 seconds after you stop typing/making changes to the document

👍 

Again, thanks a bunch for working on this!

**Comment by dometto** [comment]
> Seems sensible. @dometto do you see any reason to prefer a SHA here?

Not really, I think I was assuming we needed to use the environment templates in the page layout for this, and there we don't have access to the full URL. But this JS-only approach seems better, no need to mess with the templates. 👍

> My sense is that it would be more intuitive to recover the text by default (and maybe add an undo button).

I'm of two minds here. Doing it by default would feel nice and frictionless, but on the other hand: what if I decide to throw away my changes to a page by closing the window, and then later re-edit the page? If there was no notice, I could forget about the old changes I made and accidentally save them. Or am I missing something?
